odoo/addons/base/models/res_users.py

```diff
@@ -997,20 +997,30 @@ def create(self, vals_list):
                 values['groups_id'] = type(self).groups_id.convert_to_write(gs, user)
         return super(UsersImplied, self).create(vals_list)
 
+    def _clear_demoted_user_access(self, values):
+        users_to_demote = self.filtered(
+            lambda u: not u.has_group('base.group_user')
+        )
+        if users_to_demote:
+            vals = {'groups_id': [(5, 0, 0)] + values['groups_id']}
+            super(UsersImplied, users_to_demote).write(vals)
+        return users_to_demote
+
     def write(self, values):
+        if not values.get('groups_id'):
+            return super(UsersImplied, self).write(values)
+        # Add implied groups for all users.
         users_before = self.filtered(lambda u: u.has_group('base.group_user'))
         res = super(UsersImplied, self).write(values)
-        if values.get('groups_id'):
-            # add implied groups for all users
-            for user in self:
-                if not user.has_group('base.group_user') and user in users_before:
-                    # if we demoted a user, we strip him of all its previous privileges
-                    # (but we should not do it if we are simply adding a technical group to a portal user)
-                    vals = {'groups_id': [(5, 0, 0)] + values['groups_id']}
-                    super(UsersImplied, user).write(vals)
-                gs = set(concat(g.trans_implied_ids for g in user.groups_id))
-                vals = {'groups_id': [(4, g.id) for g in gs]}
-                super(UsersImplied, user).write(vals)
+        users_before._clear_demoted_user_access(values)
```

Comment:
I honestly don't see the added value of calling a method, to me it makes the code harder to read.  This is because the definition of "demoted" users is having the internal users group before the update and not having it after the update.  For the sake of readability, it looks simpler to me to have both calls to `filtered` in the same method.
```suggestion
        demoted_users = users_before.filtered(lambda u: not u.has_group('base.group_user'))
        if demoted_users:
            # demoted users are restricted to the assigned groups only
            vals = {'groups_id': [(5, 0, 0)] + values['groups_id']}
            super(UsersImplied, demoted_users).write(vals)
```

Reply:
Doesn't it clutter `write` method with extra stuff like handling demoted users? Also, it makes it more extend friendly, compared to when everything is in one method as you can extend/override demote users and do some extra stuff if you need to.

Reply:
I was talking about custom modules. As I always find standard Odoo code quite hard to extend, when its methods are 200+ lines long:)
Though updated as you asked.